src/replica/HttpAsyncClient.h

```diff
@@ -0,0 +1,312 @@
+    static std::string state2str(State state);
+
+    HttpAsyncClient() = delete;
+    HttpAsyncClient(HttpAsyncClient const&) = delete;
+    HttpAsyncClient& operator=(HttpAsyncClient const&) = delete;
+
+    /**
+     * @brief Static factory for creating objects of the class.
+     *
+     * Objects created by the method will be in the state State::CREATED. The actual
+     * execution of the request has to be initiated by calling the method start().
+     *
+     * @param io_service BOOST ASIO services.
+     * @param onFinish The callback function to be called upon a completion
+     *   (successfull or not) of a request.
+     * @param method The case-sensitive name of the HTTP method ('GET', 'POST', 'PUT', 'DELETE').
+     * @param url A location of the remote resoure.
+     * @param data (Optional) Data to be sent in the request's body.
+     * @param headers (Optional) HTTP headers to be send with a request.
+     * @param maxResponseBodySize (Optional) The maximum size of the response body.
+     *   If a value of the parameter is set to 0 then the default limit of 8M
+     *   imposed by the Boost.Beast library will be assumed. If the size of the response
+     *   body received from the server will exceed the (explicit or implicit) limit then
+     *   the operation will fail and the object will be put into the state State::BODY_LIMIT_ERROR.
+     * @param expirationIvalSec (Optional) A timeout to wait before the completion of a request.
+     *   The expiration timer starts after the start() method gets called. The timeout
+     *   includes all phases of the request's execution, including establishing a connection
+     *   to the server, sending the request and waiting for the server's response.
+     *   If a value of the parameter is set to 0 then no expiration timeout will be
+     *   assumed for the request.
+     * @return The shared pointer to the newly created object.
+     * @throw std::invalid_argument If empty or illegal values of the input parameters
+     *   were provided.
+     */
+    static std::shared_ptr<HttpAsyncClient> create(
+            boost::asio::io_service& io_service,
+            CallbackType const& onFinish,
+            std::string const& method,
+            std::string const& url,
+            std::string const& data=std::string(),
+            std::map<std::string, std::string> const& headers=std::map<std::string, std::string>(),
+            size_t maxResponseBodySize=0,
+            unsigned int expirationIvalSec=0);
+
+    /// Non-trivial destructor is needed to free up allocated resources.
+    virtual ~HttpAsyncClient();
+
+    /// @return The current state of the request.
+    State state() const { return _state; }
+
+    /**
+     * @brief Begin processing a request.
+     *
+     * Upon successful completion of the method the object's state will be set
+     * to State::IN_PROGRESS. Otherwise the object's state will be set to
+     * State::FAILED and an exception will be thrown.
```

Comment:
Let me explain the logic here. If a problem occurs in the constructor of the class (or its factory method `create`) then no object would be created. Once created, the object goes into the initial state `CREATED`. Both scenarios have no ambiguities. Things are getting a bit more complicated when the method `start` is attempting to initiate the request by setting up parameters of the request, initiating the expiration timer, and initiating resolve on the host's host/port. A failure to initiate the request will not automatically destroy the object (and all shared pointers to it). An exception is thrown to indicate a problem. At this point, it's not quite clear what should be done next and how to treat the object. The first possibility would be to leave the request in the state `CREATED` and assume that another call to the method `start` would "magically" resolve the issue. That's very unlikely because there is no way to "repair" the parameters of the object once it's constructed. Besides, it wouldn't be clear what exactly needs to be rapaired. The second approach that I chose here is to turn the object into the final state `FAILED` ... which also leaves the request object in an unusable state. On the surface, both are nearly the same. The only difference between those is that in the last approach the failed attempt to use the request object is now recorded in the object's state. Therefore, if this object was a part of some collection of requests, or if it was passed to some other context then the other context would be able to determine the new state of the object and take appropriate actions.

Reply:
This is fine, my question was mostly that it's not clear what can cause this method to fail, as I thought that most checks and hostname/port parsing/resolution should be done in constructor.

Reply:
I will fix the documentation on the method along with pushing parameter validation to the c-tor.